File: sliding window (SubArray)/Largest subarray with 0 sum/Largest_subarray_with_0_sum.py
#Your task is to complete this function
 
class Solution:
    #Approach:Brute Force
    #Time Complexity:O(n^2)
    #Space Complexity:O(1)
    def maxLen(self, arr):
        maxi=0
        for i in range(0,len(arr)):
            sum=0
            for j in range(i,len(arr)):
                sum+=arr[j]
                if(sum==0):
                    maxi=max(maxi,j-i+1)
        return maxi
    
    #Approach:Optimal
    #Time Complexity:O(n)
    #Space Complexity:O(n)
    def maxLen_1(self, arr):
        maxi=0
        sum=0
        Map={0:-1}
        
        for i in range(0,len(arr)):
            sum+=arr[i]
            if(sum==0):
                diff=i-Map[sum]
                maxi=max(maxi,diff)
            elif Map.get(sum) is not None:
                diff=i-Map[sum]
                maxi=max(maxi,diff)
            else:
                Map.update({sum:i})
        return maxi
        

if __name__=='__main__':
    arr=[15, -2, 2, -8, 1, 7, 10, 23,0,0,0,0,0,0,0]
    ob=Solution()
    print(ob.maxLen_1(arr))

File: sliding window (SubArray)/Largest subarray with 0 sum/Largest_subarray_with_0_sum.java

import java.util.HashMap;

class Solution {

    // Brute force solution
    // Time complexity O(n^2)
    // Space complexity O(1)
    int maxLen_1(int arr[]) {
        int n = arr.length;
        int maxi = 0;
        for (int i = 0; i < n; i++) {
            int sum = 0;
            for (int j = i; j < n; j++) {
                sum += arr[j];
                if (sum == 0) {
                    maxi = Math.max(maxi, j - i + 1);
                }
            }
        }
        return maxi;
    }

    // Time complexity O(n)
    // Space complexity O(n)
    //Approach: 1. We will use a hashmap to store the sum and its index.
    int maxLen(int arr[]) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int maxi = 0;
        int sum = 0;
        map.put(0, -1);
        for (int i = 0; i < arr.length; i++) {
            sum += arr[i];
            if (sum == 0) {
                int diff = i - map.get(sum);
                maxi = Math.max(maxi, diff);
            } else if (map.containsKey(sum)) {
                int diff = i - map.get(sum);
                maxi = Math.max(maxi, diff);
            } else {
                map.put(sum, i);
            }
        }
        return maxi;
    }
}

public class Largest_subarray_with_0_sum {

    public static void main(String[] args) {
        Solution sol = new Solution();
        int arr[] = {15, -2, 2, -8, 1, 7, 10, 23};
        int ans = sol.maxLen(arr);
        int ans_1 = sol.maxLen_1(arr);
        System.out.println(ans);
        System.out.println(ans_1);
    }
}
